espnet/nets/pytorch_backend/e2e_transformer.py

```diff
@@ -0,0 +1,734 @@
+            )
+        elif args.transformer_input_layer == "conv2d":
+            self.input_layer = Conv2dSubsampling(idim, args.adim, args.dropout_rate)
+        elif args.transformer_input_layer == "embed":
+            self.input_layer = torch.nn.Sequential(
+                torch.nn.Embedding(idim, args.adim),
+                PositionalEncoding(args.adim, args.dropout_rate)
+            )
+        else:
+            raise ValueError("unknown input_layer: " + args.transformer_input_layer)
+
+        self.encoders = repeat(
+            args.elayers,
+            lambda: EncoderLayer(
+                args.adim,
+                MultiHeadedAttention(args.aheads, args.adim, args.transformer_attn_dropout_rate),
+                PositionwiseFeedForward(args.adim, args.eunits, args.dropout_rate),
+                args.dropout_rate
+            )
+        )
+        self.norm = LayerNorm(args.adim)
+
+    def forward(self, x, mask):
+        if isinstance(self.input_layer, Conv2dSubsampling):
+            x, mask = self.input_layer(x, mask)
+        else:
+            x = self.input_layer(x)
+        x, mask = self.encoders(x, mask)
+        return self.norm(x), mask
+
+
+class Decoder(torch.nn.Module):
+    def __init__(self, odim, args):
+        super(Decoder, self).__init__()
+        self.embed = torch.nn.Sequential(
+            torch.nn.Embedding(odim, args.adim),
+            PositionalEncoding(args.adim, args.dropout_rate)
+        )
+        self.decoders = repeat(
+            args.dlayers,
+            lambda: DecoderLayer(
+                args.adim,
+                MultiHeadedAttention(args.aheads, args.adim, args.transformer_attn_dropout_rate),
+                MultiHeadedAttention(args.aheads, args.adim, args.transformer_attn_dropout_rate),
+                PositionwiseFeedForward(args.adim, args.dunits, args.dropout_rate),
+                args.dropout_rate
+            )
+        )
+        self.output_norm = LayerNorm(args.adim)
+        self.output_layer = torch.nn.Linear(args.adim, odim)
+
+    def forward(self, tgt, tgt_mask, memory, memory_mask):
+        """forward decoder
+
+        :param torch.Tensor tgt: input token ids, int64 (batch, maxlen_out)
+        :param torch.Tensor tgt_mask: input token mask, uint8  (batch, maxlen_out)
+        :param torch.Tensor memory_mask: encoded memory, float32  (batch, maxlen_in, feat)
+        :param torch.Tensor memory_mask: encoded memory mask, uint8  (batch, maxlen_in)
+        :return x: decoded token score before softmax (batch, maxlen_out, token)
+        :rtype: torch.Tensor
+        :return tgt_mask: score mask before softmax (batch, maxlen_out)
+        :rtype: torch.Tensor
+        """
+        x = self.embed(tgt)
+        x, tgt_mask, memory, memory_mask = self.decoders(x, tgt_mask, memory, memory_mask)
+        x = self.output_layer(self.output_norm(x))
+        return x, tgt_mask
+
+
+class LabelSmoothing(nn.Module):
```

Comment:
Is there any difference between @creatorscan and this label smoothing?
It would be better to unify it in future.
@creatorscan, could you also check this?

Reply:
This LS implementation is originally based on this paper https://arxiv.org/pdf/1512.00567v1.pdf as described in transformer paper and opennmt-py.

There are two differences from the one in ESPnet.
1. it masks padded tokens (it is not blank=0 but ignore_id=-1 in the target tensor)
2. it is uniform smoothing (1.0 - w or w) instead of unigram histgram

Reply:
In my opinion, it is non sense to unify this because its mechanism is completely different (but its motivation is same). Maybe I will change the name of argparse option instead of reusing it. It should be able to use at the same time e.g., `--unigram_lsm 0.1 --uniform_lsm 0.1`.

Reply:
I see. Year, it's better to avoid to reuse the option then.